### PR TITLE
Reuse connection when calling ReadToCount()

### PR DESCRIPTION
**Pull request by vhvb1989**
Fixing the case when customer knows the size of the downloaded blob stream and calls ReadToCount(..) asking for the specific size to be pulled.
In this case, the EOF flag was not updated at the end of reading all because it it updated until user try to read again.

Fixing this by removing EOF state and adding IsEOF function to check both chunked and non-chunked responses.

Fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/446

### CODE REVIEW

File: sdk/core/azure-core/inc/http/curl/curl.hpp [hunk 1]
```diff
@@ -401,17 +394,17 @@ namespace Azure { namespace Core { namespace Http {
...
+      this->m_sessionTotalRead = 0;
     }
 
     ~CurlSession() override
     {
       // mark connection as reusable only if entire response was read
       // If not, connection can't be reused because next Read will start from what it is currently
       // in the wire. We leave the connection blocked until Server closes the connection
-      if (this->m_rawResponseEOF)
+      if (IsEOF())
```
Comment (JeffreyRichter) [inline sdk/core/azure-core/inc/http/curl/curl.hpp:407]: Is this really the best thing to do? Maybe we should read to the end and then put the connection back in the pool. Is there ANY benefit to just letting the connection linger?

File: sdk/core/azure-core/inc/http/curl/curl.hpp [hunk 2]
```diff
@@ -437,6 +430,12 @@ namespace Azure { namespace Core { namespace Http {
     int64_t Length() const override { return this->m_contentLength; }
 
     int64_t Read(Azure::Core::Context const& context, uint8_t* buffer, int64_t count) override;
+
+    bool IsEOF()
```
Comment (JeffreyRichter) [inline sdk/core/azure-core/inc/http/curl/curl.hpp:434]: This is not public, right?

File: sdk/core/azure-core/inc/http/curl/curl.hpp (hunk 1 above)
Comment (JinmingHu-MSFT) [reply sdk/core/azure-core/inc/http/curl/curl.hpp:407]: it's gonna take some time if the body to read is huge. In this case we might as well close this connection.

**Comment by JinmingHu-MSFT** [comment]
@vhvb1989 This pr introduces a hang issue

**Review by JinmingHu-MSFT — CHANGES_REQUESTED** [review]
introduces a hang issue, don't merge before fixing

**Comment by vhvb1989** [comment]
> introduces a hang issue, don't merge before fixing

Ci passed. Where do you see the hang?

**Comment by JinmingHu-MSFT** [comment]
> > introduces a hang issue, don't merge before fixing
> 
> Ci passed. Where do you see the hang?

i'm going to give you a small program that can reproduce.

**Comment by JinmingHu-MSFT** [comment]
> > > introduces a hang issue, don't merge before fixing
> > 
> > 
> > Ci passed. Where do you see the hang?
> 
> i'm going to give you a small program that can reproduce.

sent you over Teams

**Review by JinmingHu-MSFT — APPROVED** [review]
storage test cases all passed.

File: sdk/core/azure-core/inc/http/curl/curl.hpp (hunk 1 above)
Comment (vhvb1989) [reply sdk/core/azure-core/inc/http/curl/curl.hpp:407]: yeah, we can't read to end as @JinmingHu-MSFT  said.

Note: We **don't** let connection linger.  I will update the comment on it.
As soon as the connection is not moved back to the pool, it is destroyed.  It's destructor will clean libcurl handle, so the connection is closed.

File: sdk/core/azure-core/inc/http/curl/curl.hpp (hunk 2 above)
Comment (vhvb1989) [reply sdk/core/azure-core/inc/http/curl/curl.hpp:434]: it was. I just moved to be private to the Session only

**Pull request merged**